### PR TITLE
feat: add close method for the region trait

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Adds `close` method for `Region` trait and implements the`close` method for the corresponding structs. Therefore, the upper(e.g. `EngineInner`) can call the `close` method to release the region resources when closing or dropping regions.

Based on my understanding, in order to close a region,

1. (Frontend) The frontend sets a CLOSING flag for the target region in metasrv. After that, it will reject future writing / altering requests.
2. (Frontend) Sends the close request to datanode leader(s).
 - [x] (Datanode-leader) Leader acquires the write lock.
- [x] (Datanode-leader) Sets a memory flag to reject any potential(due to network delay) writing.
- [x] (Datanode-leader) Waits for the pending flush task.
- [ ] (Datanode-leader) Cancels compaction task (TODO #930 )
3. (Frontend) The frontend sets a CLOSED flag for the target region in metasrv.

Untitled Diagram (5)
(rejecting the write requests due to network delay)

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

#832

### CODE REVIEW

**Comment by WenyXu** [comment]
> > 1. (Frontend) The frontend sets a CLOSING flag for the target region in metasrv. After that, it will reject future writing / altering requests.
> 
> We could focus on how to close a region on datanode first. Then we can implement a method to shut down the whole storage engine (and the datanode) gracefully.

I see. It seems we need to listen to the os.exit signal, and if the signal was received, then close the storage engine, which includes closing all regions stored in the datanode gracefully.

File: src/storage/src/region/writer.rs [hunk 1]
```diff
@@ -220,6 +226,59 @@ impl RegionWriter {
...
+
+            if inner.is_closed() {
+                return Ok(());
+            }
+
+            inner.close();
+        }
+        // we release the writer lock once for rejecting any following potential writing requests immediately.
+
+        self.trigger_all_flush(writer_ctx).await?;
```
Comment (evenyag) [inline src/storage/src/region/writer.rs:253]: While closing, we should cancel all background jobs to avoid blocking on flush or compaction for a long time.

Comment (WenyXu) [reply src/storage/src/region/writer.rs:253]: I got you. We could recover the pending writes from the WAL.

**Comment by WenyXu** [comment]
> > 1. (Frontend) The frontend sets a CLOSING flag for the target region in metasrv. After that, it will reject future writing / altering requests.
> 
> We could focus on how to close a region on datanode first. Then we can implement a method to shut down the whole storage engine (and the datanode) gracefully.

By the way, should I implement the close method for MitoTable & MitoEngine to handle graceful shutdown(i.e. handle the closing while receiving the SIGTERM / SIGINT)?  or leaf those to another issue?

**Comment by evenyag** [comment]
> By the way, should I implement the close method for MitoTable & MitoEngine to handle graceful shutdown

Yes, we should provide a close method.

(i.e. handle the closing while receiving the SIGTERM / SIGINT)? or leaf those to another issue?

Both `storage` and `mito` are libraries so they don't need to care about signal handling. We handle the signal in the application
https://github.com/GreptimeTeam/greptimedb/blob/8ffc078f88f5e265eb0821820593a1c4bd309c17/src/cmd/src/bin/greptime.rs#L105-L107

We should shut down all components after we receive the `ctrl-c` signal. If you still have questions, we could discuss them in the [issue](https://github.com/GreptimeTeam/greptimedb/issues/832) or open a new issue for that.

File: src/storage/src/region/writer.rs [hunk 2]
```diff
@@ -220,6 +226,56 @@ impl RegionWriter {
...
+
+        self.cancel_flush().await?;
+
+        // TODO: canncel the compaction task
+
+        Ok(())
+    }
+
+    /// Cancel flush task if any
+    pub async fn cancel_flush(&self) -> Result<()> {
```
Comment (evenyag) [inline src/storage/src/region/writer.rs:254]: Do we need to pub this method?

File: src/storage/src/region/writer.rs [hunk 3]
```diff
@@ -220,6 +226,56 @@ impl RegionWriter {
...
+                .await
+                .map_err(BoxedError::new)
+                .context(error::CancelFlushSnafu)?;
+        }
+
+        Ok(())
+    }
+
+    /// Trigger the flush task manually.
+    pub async fn trigger_flush<S: LogStore>(&self, writer_ctx: WriterContext<'_, S>) -> Result<()> {
```
Comment (evenyag) [inline src/storage/src/region/writer.rs:268]: Do we still need it in this PR?

File: src/storage/src/region/writer.rs [hunk 4]
```diff
@@ -233,6 +289,13 @@ impl RegionWriter {
 
         Ok(())
     }
+
+    pub async fn set_flush_task(&self, task: JobHandle) -> Result<()> {
```
Comment (evenyag) [inline src/storage/src/region/writer.rs:293]: Do we still need it in this PR?

File: src/datanode/src/datanode.rs [hunk 5]
```diff
@@ -175,4 +175,18 @@ impl Datanode {
     pub fn get_instance(&self) -> InstanceRef {
         self.instance.clone()
     }
+
+    pub async fn shutdown_instance(&self) -> Result<()> {
```
Comment (evenyag) [inline src/datanode/src/datanode.rs:179]: As the main purpose of this PR is adding close method to the region, I'd suggest revert the changes introduced by [387ada](https://github.com/GreptimeTeam/greptimedb/pull/970/commits/387ada6c54380cfac833a7415f508f00afd1f7e5)

Comment (WenyXu) [reply src/datanode/src/datanode.rs:179]: > As the main purpose of this PR is adding close method to the region, I'd suggest revert the changes introduced by [387ada](https://github.com/GreptimeTeam/greptimedb/pull/970/commits/387ada6c54380cfac833a7415f508f00afd1f7e5)

I see, let me handle it.

File: src/storage/src/region/writer.rs [hunk 6]
```diff
@@ -617,4 +666,13 @@ impl WriterInner {
         });
         Some(schedule_compaction_cb)
     }
+
+    #[inline]
+    fn is_closed(&self) -> bool {
+        self.closed
+    }
+
+    fn mark_closed(&mut self) {
```
Comment (v0y4g3r) [inline src/storage/src/region/writer.rs:675]: Please also inline this method.

File: src/storage/src/error.rs [hunk 7]
```diff
@@ -141,6 +141,12 @@ pub enum Error {
     #[snafu(display("Task already cancelled"))]
     Cancelled { backtrace: Backtrace },
 
+    #[snafu(display("Failed to cancel flush, source: {}", source))]
+    CancelFlush {
+        source: BoxedError,
+        backtrace: Backtrace,
+    },
```
Comment (v0y4g3r) [inline src/storage/src/error.rs:148]:
```suggestion
    #[snafu(display("Failed to cancel flush, source: {}", source))]
    CancelFlush {
        #[snafu(backtrace)]
        source: BoxedError,
    },
```

File: src/storage/src/error.rs [hunk 8]
```diff
@@ -232,6 +245,9 @@ pub enum Error {
         source: MetadataError,
     },
 
+    #[snafu(display("Try to write the closed region"))]
+    ClosedRegion {},
```
Comment (v0y4g3r) [inline src/storage/src/error.rs:249]: Should we add a backtrace here?

File: src/storage/src/wal.rs [hunk 9]
```diff
@@ -78,10 +78,14 @@ impl<S: LogStore> Wal<S> {
         self.region_id
     }
 
-    #[cfg(test)]
     pub async fn close(&self) -> Result<()> {
-        let _ = self.store.stop().await;
-        Ok(())
+        self.store
```
Comment (v0y4g3r) [inline src/storage/src/wal.rs:82]: correctness: we should not stop logstore here. `Wal` only holds a reference to `Logstore`, it does not actually own it.   Otherwise closing a region will stop the logstore, which will make the whole storage engine malfunction.

**Review by v0y4g3r — CHANGES_REQUESTED** [review]

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Review by evenyag — APPROVED** [review]
LGTM
